Refactor native compilation

ksew1 opened the pull request:
**Stack**:
- #3711
- #3710
- #3709 ⬅
- #3713


⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

ksew1 committed:
Refactor native compilation

commit-id:26525dd3

cptartur commented on crates/native-api/src/lib.rs:
```diff
@@ -0,0 +1,40 @@
+    AotContractExecutor::new(
+        &sierra_program,
+        &contract_class.entry_points_by_type,
+        sierra_version.clone().into(),
```
I don't think we officially support any scarb version that produces incompatible sierra, but as a sanity check, maybe we should validate the version somewhere.

I think native only supports sierra from 1.7.0 upwards.

Maybe minimal scarb version in foundry should be bumped to lowest scarb that supports that version? (Still there's a problem of contracts from network in forking, we should ensure we don't compile incompatible ones).

cptartur commented on crates/scarb-api/src/artifacts.rs:
```diff
@@ -12,34 +13,53 @@ mod deserialized;
 
 /// Contains compiled Starknet artifacts
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, Clone)]
 pub struct StarknetContractArtifacts {
     /// Compiled sierra code
     pub sierra: String,
     /// Compiled casm code
     pub casm: String,
+
+    /// Optional AOT compiled native executor
+    pub executor: Option<AotContractExecutor>,
+}
```
To be 100% idiomatic, I think this struct should be moved to forge crate or somewhere else so that that impl is owned by the user. It doesn't make much sense for Scarb api to own a struct for `StarknetContractArtifacts` with details about native compilation if scarb knows nothing about that.

cptartur replied:
And while we are at it, having casm here if compilation is done by USC not scarb is also weird.

cptartur replied:
Unless refactoring this is an extremely big lift, I think now might be a good moment to finally untangle that.

cptartur reviewed (approved):
Probably disregard bigger refactoring for now

ksew1 committed:
Add CLI for cairo native (#3710)

Closes #3689

---

**Stack**:
- #3711
- #3710 ⬅
- #3709
- #3713


⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do
not merge manually using the UI - doing so may have unexpected results.*
